File: examples/java/CacheExample.java
/**
 * FastCollection v1.0.0 - Cache Example
 * 
 * This example demonstrates using FastCollectionMap as a cache with TTL.
 * 
 * Copyright © 2025-2030, Ashutosh Sinha (ajsinha@gmail.com)
 * Patent Pending
 */
package examples;

import com.kuber.fastcollection.*;
import java.io.Serializable;
import java.util.*;

public class CacheExample {
    
    /**
     * Simple cache wrapper around FastCollectionMap
     */
    public static class Cache<K extends Serializable, V extends Serializable> {
        private final FastCollectionMap<K, V> store;
        private final int defaultTTL;
        
        public Cache(String path, int defaultTTLSeconds) {
            this.store = new FastCollectionMap<>(path, 32 * 1024 * 1024, true);
            this.defaultTTL = defaultTTLSeconds;
        }
        
        public void put(K key, V value) {
            store.put(key, value, defaultTTL);
        }
        
        public void put(K key, V value, int ttlSeconds) {
            store.put(key, value, ttlSeconds);
        }
        
        public V get(K key) {
            return store.get(key);
        }
        
        public V getOrDefault(K key, V defaultValue) {
            V value = store.get(key);
            return value != null ? value : defaultValue;
        }
        
        public boolean contains(K key) {
            return store.containsKey(key);
        }
        
        public void remove(K key) {
            store.remove(key);
        }
        
        public long getTTL(K key) {
            return store.getTTL(key);
        }
        
        public void refreshTTL(K key) {
            store.setTTL(key, defaultTTL);
        }
        
        public int size() {
            return store.size();
        }
        
        public int cleanup() {
            return store.removeExpired();
        }
        
        public void close() {
            store.close();
        }
    }
    
    public static void main(String[] args) throws InterruptedException {
        System.out.println("FastCollection v1.0.0 - Cache Example");
        System.out.println("=====================================\n");
        
        // Create cache with 10-second default TTL
        Cache<String, String> cache = new Cache<>("/tmp/cache_example.fc", 10);
        
        try {
            // Store user data
            System.out.println("Storing user data...");
            cache.put("user:1001", "John Doe");
            cache.put("user:1002", "Jane Smith");
            cache.put("user:1003", "Bob Johnson");
            
            // Store with custom TTL
            cache.put("session:abc123", "session_data", 30);  // 30 seconds
            cache.put("config:app", "config_value", -1);      // Never expires
            
            System.out.println("Cache size: " + cache.size());
            
            // Retrieve data
            System.out.println("\nRetrieving data:");
            System.out.println("  user:1001 = " + cache.get("user:1001"));
            System.out.println("  user:1002 = " + cache.get("user:1002"));
            System.out.println("  nonexistent = " + cache.getOrDefault("nonexistent", "DEFAULT"));
            
            // Check TTLs
            System.out.println("\nTTL values:");
            System.out.println("  user:1001 TTL = " + cache.getTTL("user:1001") + "s");
            System.out.println("  session:abc123 TTL = " + cache.getTTL("session:abc123") + "s");
            System.out.println("  config:app TTL = " + cache.getTTL("config:app") + " (infinite)");
            
            // Wait for some items to expire
            System.out.println("\nWaiting 12 seconds for user data to expire...");
            Thread.sleep(12000);
            
            // Check what's left
            System.out.println("\nAfter expiry:");
            System.out.println("  Cache size: " + cache.size());
            System.out.println("  user:1001 = " + cache.get("user:1001"));  // null (expired)
            System.out.println("  session:abc123 = " + cache.get("session:abc123"));  // still valid
            System.out.println("  config:app = " + cache.get("config:app"));  // never expires
            
            // Cleanup expired entries
            int removed = cache.cleanup();
            System.out.println("\nCleaned up " + removed + " expired entries");
            
            System.out.println("\nExample completed successfully!");
            
        } finally {
            cache.close();
        }
    }
}

File: examples/python/cache_example.py
#!/usr/bin/env python3
"""
FastCollection v1.0.0 - Cache Example (Python)

This example demonstrates using FastMap as a cache with TTL.

Usage: python cache_example.py

Copyright © 2025-2030, Ashutosh Sinha (ajsinha@gmail.com)
Patent Pending
"""

from fastcollection import FastMap
import json
import time


class Cache:
    """Simple cache wrapper around FastMap"""
    
    def __init__(self, path: str, default_ttl: int = 300):
        self.store = FastMap(path)
        self.default_ttl = default_ttl
    
    def put(self, key: str, value, ttl: int = None):
        """Store a value with optional TTL"""
        ttl = ttl if ttl is not None else self.default_ttl
        data = json.dumps(value).encode()
        self.store.put(key.encode(), data, ttl=ttl)
    
    def get(self, key: str, default=None):
        """Retrieve a value"""
        data = self.store.get(key.encode())
        if data is None:
            return default
        return json.loads(data.decode())
    
    def contains(self, key: str) -> bool:
        """Check if key exists"""
        return self.store.contains_key(key.encode())
    
    def delete(self, key: str):
        """Delete a key"""
        self.store.remove(key.encode())
    
    def get_ttl(self, key: str) -> int:
        """Get remaining TTL in seconds"""
        return self.store.get_ttl(key.encode())
    
    def refresh_ttl(self, key: str, ttl: int = None):
        """Refresh TTL for a key"""
        ttl = ttl if ttl is not None else self.default_ttl
        self.store.set_ttl(key.encode(), ttl)
    
    def size(self) -> int:
        """Get number of items"""
        return self.store.size()
    
    def cleanup(self) -> int:
        """Remove expired items"""
        return self.store.remove_expired()
    
    def close(self):
        """Close the cache"""
        self.store.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, *args):
        self.close()


def main():
    print("FastCollection v1.0.0 - Cache Example (Python)")
    print("=" * 50)
    print()
    
    # Create cache with 10-second default TTL
    with Cache("/tmp/cache_example_py.fc", default_ttl=10) as cache:
        # Store user data
        print("Storing user data...")
        cache.put("user:1001", {"name": "John Doe", "email": "john@example.com"})
        cache.put("user:1002", {"name": "Jane Smith", "email": "jane@example.com"})
        cache.put("user:1003", {"name": "Bob Johnson", "email": "bob@example.com"})
        
        # Store with custom TTL
        cache.put("session:abc123", {"token": "xyz", "user_id": "1001"}, ttl=30)
        cache.put("config:app", {"debug": True, "version": "1.0.0"}, ttl=-1)  # Never expires
        
        print(f"Cache size: {cache.size()}")
        
        # Retrieve data
        print("\nRetrieving data:")
        user1 = cache.get("user:1001")
        print(f"  user:1001 = {user1}")
        
        user2 = cache.get("user:1002")
        print(f"  user:1002 = {user2}")
        
        default = cache.get("nonexistent", {"default": True})
        print(f"  nonexistent = {default}")
        
        # Check TTLs
        print("\nTTL values:")
        print(f"  user:1001 TTL = {cache.get_ttl('user:1001')}s")
        print(f"  session:abc123 TTL = {cache.get_ttl('session:abc123')}s")
        print(f"  config:app TTL = {cache.get_ttl('config:app')} (infinite)")
        
        # Wait for some items to expire
        print("\nWaiting 12 seconds for user data to expire...")
        time.sleep(12)
        
        # Check what's left
        print("\nAfter expiry:")
        print(f"  Cache size: {cache.size()}")
        print(f"  user:1001 = {cache.get('user:1001')}")  # None (expired)
        print(f"  session:abc123 = {cache.get('session:abc123')}")  # still valid
        print(f"  config:app = {cache.get('config:app')}")  # never expires
        
        # Cleanup expired entries
        removed = cache.cleanup()
        print(f"\nCleaned up {removed} expired entries")
        
        print("\nExample completed successfully!")


if __name__ == "__main__":
    main()
